crates/vm/levm/src/memory.rs

```diff
@@ -279,23 +278,20 @@ pub fn expansion_cost(new_memory_size: usize, current_memory_size: usize) -> Res
 }
 
 /// The total cost for a given memory size.
+/// Gas cost should always be computed in u64
 #[inline]
 fn cost(memory_size: usize) -> Result<u64, VMError> {
-    let memory_size_word = memory_size
-        .checked_add(WORD_SIZE_IN_BYTES_USIZE.wrapping_sub(1))
-        .ok_or(OutOfGas)?
-        / WORD_SIZE_IN_BYTES_USIZE;
-
-    let gas_cost = (memory_size_word
-        .checked_mul(memory_size_word)
-        .ok_or(OutOfGas)?
-        / MEMORY_EXPANSION_QUOTIENT)
-        .checked_add(3usize.checked_mul(memory_size_word).ok_or(OutOfGas)?)
-        .ok_or(OutOfGas)?;
-
-    gas_cost
-        .try_into()
-        .map_err(|_| ExceptionalHalt::VeryLargeNumber.into())
+    let memory_size = u64::try_from(memory_size).map_err(|_| InternalError::TypeConversion)?;
+
+    // memory size measured in 32 byte words
+    let words = memory_size.div_ceil(WORD_SIZE_IN_BYTES_U64);
+
+    // Cost(words) ≈ floor(words^2 / q) + 3 * words
+    // For this to overflow memory size in words should be 2^32, which is impossible.
```

Comment:
What makes it impossible? I would guess our implementation bounds it, but the comment doesn't make that explicit.
Can we add a `debug_assert` to detect illegal behavior in debug builds?